File: packages/kt-client/src/main/kotlin/dev/pyric/database/MutableData.kt
package dev.pyric.database

class MutableData(
    var value: Any?,
    var priority: Any? = null,
    val key: String? = null
) {
    val childrenCount: Long
        get() = when (val v = value) {
            is Map<*, *> -> v.size.toLong()
            is List<*> -> v.count { it != null }.toLong()
            else -> 0L
        }

    val children: Iterable<MutableData>
        get() = when (val v = value) {
            is Map<*, *> -> v.entries.map { (k, childVal) ->
                MutableData(childVal, null, k.toString())
            }
            is List<*> -> v.mapIndexedNotNull { idx, item ->
                if (item == null) null else MutableData(item, null, idx.toString())
            }
            else -> emptyList()
        }

    fun hasChildren(): Boolean = childrenCount > 0L

    fun hasChild(path: String): Boolean = child(path).value != null

    fun child(path: String): MutableData {
        val segments = path.trim('/').split('/').filter { it.isNotEmpty() }
        if (segments.isEmpty()) return this

        var currentVal = value
        for (segment in segments) {
            currentVal = when (val v = currentVal) {
                is Map<*, *> -> v[segment]
                is List<*> -> {
                    val idx = segment.toIntOrNull()
                    if (idx != null && idx in v.indices) v[idx] else null
                }
                else -> null
            }
        }
        return MutableData(currentVal, null, segments.last())
    }
}

File: packages/swift-client/Sources/PyricDatabase/MutableData.swift
import Foundation
import PyricFirestore

/// A mutable snapshot of data at a Realtime Database location used in transactions.
public final class MutableData: NSObject, @unchecked Sendable {
    public var value: Any?
    public var priority: Any?
    public let key: String?

    public init(key: String? = nil, value: Any? = nil, priority: Any? = nil) {
        self.key = key
        self.value = value
        self.priority = priority
        super.init()
    }

    public var childrenCount: UInt {
        if let dict = value as? [String: Any] {
            return UInt(dict.count)
        }
        return 0
    }

    public func hasChildren() -> Bool {
        return childrenCount > 0
    }

    public func hasChild(atPath path: String) -> Bool {
        let child = childData(byAppendingPath: path)
        return child.value != nil && !(child.value is NSNull)
    }

    public func childData(byAppendingPath path: String) -> MutableData {
        let segments = path.split(separator: "/").map(String.init).filter { !$0.isEmpty }
        guard !segments.isEmpty else { return self }

        var curVal = value
        for seg in segments {
            if let dict = curVal as? [String: Any] {
                curVal = dict[seg]
            } else {
                curVal = nil
            }
        }
        return MutableData(key: segments.last, value: curVal, priority: nil)
    }

    public var children: NSEnumerator {
        guard let dict = value as? [String: Any] else {
            return ([] as NSArray).objectEnumerator()
        }
        let sortedKeys = dict.keys.sorted()
        let items = sortedKeys.map { k in
            MutableData(key: k, value: dict[k], priority: nil)
        }
        return (items as NSArray).objectEnumerator()
    }
}

/// The result of a Realtime Database transaction block.
public final class TransactionResult: NSObject, @unchecked Sendable {
    public let isAborted: Bool
    public let mutableData: MutableData?

    private init(isAborted: Bool, mutableData: MutableData?) {
        self.isAborted = isAborted
        self.mutableData = mutableData
        super.init()
    }

    /// Commits the transaction with the updated MutableData value.
    public static func success(withValue value: MutableData) -> TransactionResult {
        return TransactionResult(isAborted: false, mutableData: value)
    }

    /// Aborts the transaction without saving any changes.
    public static func abort() -> TransactionResult {
        return TransactionResult(isAborted: true, mutableData: nil)
    }
}
